### ISSUE TITLE
[BUG] Linux: Unable to load shared library 'SDL2_ttf.dll'

### ISSUE THREAD

**Issue by SuperTux88** [ShadowTheAge/yafc#22]
Version 0.4.2 on Linux (but I already had the same problem with 0.4.1)

When trying to select text in an input field (for example when searching in the dependency explorer, type something in the search field, then trying to select the text with the mouse), this error shows up:

![image](https://user-images.githubusercontent.com/458548/82762064-1f906700-9dff-11ea-97a7-d6fedb67d5a6.png)

The `SDL2_ttf.dll` exists and is in the zip (but looks like it's a Windows DLL?), but there is no `libSDL2_ttf.dll` (so the `No such file or directory` in the error is correct). I installed all SDL2 dependencies listed in the [installation instructions](https://github.com/ShadowTheAge/yafc/blob/master/Docs/LinuxOsxInstall.md) (so `libsdl2-ttf` is installed, because without it, it wouldn't even start).

I then set the LD_DEBUG environment variable as written in the error message and I can see that it first tries to access `SDL2_ttf.dll.so`, then `libSDL2_ttf.dll.so`, then `SDL2_ttf.dll` and finally `libSDL2_ttf.dll` which then fails and is mentioned in the error message because it doesn't exist. And it only searched in `/lib64/` and `/usr/lib64/` (which is ok, because that's where my SDL2 libraries are installed, but on Linux they aren't DLLs, the real file it should load is `libSDL2_ttf.so`)

If I copy the existing `SDL2_ttf.dll` to `/usr/lib64/libSDL2_ttf.dll` where it then finds it, it still shows a red error, but the message is now `/usr/lib64/libSDL2_ttf.dll: invalid ELF header`, because as said, it looks like the DLLs included in the `Linux.zip` are Windows DLLs.

If I symlink my installed `/usr/lib64/libSDL2_ttf.so` to `/usr/lib64/libSDL2_ttf.dll.so` where it searches for it, it loads it correctly and I can select text in input fields without an error. But I don't understand why it only tries to load it with `.dll` in the name (especially since it already loads the correct `.so` file on startup).

**Comment by ShadowTheAge** [comment]
Thats very strange because the fact that it displays error message with *text* clearly means that ttf library is loaded and working

**Comment by ShadowTheAge** [comment]
Aha, found a problem
Library names are remapped in linux, but there is one more place that I forgot to remap

**Comment by SuperTux88** [comment]
Oh, I just found the remapping code (and the `SDL2_ttf.dll` is in there, which explains why it's correctly loaded at the start and text is working):

https://github.com/ShadowTheAge/yafc/blob/fd836c4df94dc764b5a7366d295d9fa1e7bd988d/YAFC/Program.cs#L36-L46

That also explains why I needed to create a symlink `SDL2-2.0.so` -> `libSDL2-2.0.so.0`, but I wasn't sure if that's an issue with the packages of my distribution (Gentoo), or not (so I didn't create an issue for that). The other two SDL libs are mapped to `*.so.0`, but the `SDL2-2.0.so` is mapped without the `.0` at the end. All 3 SDL libs only exist with `.0` at the end in my distribution, did you have a specific reason to map `SDL2-2.0.so` without the `.0` at the end? (not related to the `SDL2_ttf.dll` problem above, but since you mentioned the mapping I thought I'd ask)

**Comment by ShadowTheAge** [comment]
well, because in the distribution I have tried to compile for (Debian), the SDL shared object is named like in that code snippet

**Comment by SuperTux88** [comment]
Interesting, the debian package also [doesn't contain](https://packages.debian.org/buster/amd64/libsdl2-2.0-0/filelist) the `libSDL2-2.0.so`. After installing all 3 SDL2 packages, only these files and symlinks exist on debian:

```
/usr/lib/x86_64-linux-gnu/libSDL2-2.0.so.0 -> libSDL2-2.0.so.0.9.0
/usr/lib/x86_64-linux-gnu/libSDL2-2.0.so.0.9.0
/usr/lib/x86_64-linux-gnu/libSDL2_image-2.0.so.0 -> libSDL2_image-2.0.so.0.2.2
/usr/lib/x86_64-linux-gnu/libSDL2_image-2.0.so.0.2.2
/usr/lib/x86_64-linux-gnu/libSDL2_ttf-2.0.so.0 -> libSDL2_ttf-2.0.so.0.14.1
/usr/lib/x86_64-linux-gnu/libSDL2_ttf-2.0.so.0.14.1
```

Strange that it still worked for you, but it would probably be better if you also add the `.0` to the mapping for `SDL2-2.0.so`

**Comment by ShadowTheAge** [comment]
okay, did it (I will make a release tomorrow)
I think I will also add these sdl bindings in the package. Do Linux shared objects usually work between distributions?

**Comment by SuperTux88** [comment]
> Do Linux shared objects usually work between distributions?

It's complicated, but the files in the Debian packages are linked against the other libraries shipped with Debian, so they work with Debian (or to be exact, with that version of Debian), and might work on other distros/versions, but it will probably break (I tried to build the current git version and with the included SDL shared objects it just segfaults on my system, deleting them lets me start it again). You could bundle static linked versions of the libs, but you would then need to build, maintain and update the libs in your bundle, and that's probably nothing you want to do. So it's way easier to just use the SDL the distributions include.

So Debian has the `.0` at the end of the SDL libs everywhere, and on my system with Gentoo as well, Ubuntu and Mint and all other derivates of Debian have the same package structure (maybe different versions, but the `libSDL2-2.0.so.0` always symlinks to the actual installed version). I also checked Fedora, which is another big Distribution based on Redhat, and there it's the same filenames as well (and I guess all others based on that have the same structure as well). and With that you probably already have over 90% most linux users. So just adding the `.0` to libSDL (as you already changed it), should cover most of the users. So just remove the included shared objects again, because they probably break on a lot of systems.

> Or you can test it now if you pull master using dotnet publish in the source folder

As already mentioned above, I did that. And after deleting the now bundled SDL shared objects, it at least lets me start it, and I get to the Welcome screen. But then trying to load it, crashes again (also trying to create a new file and not loading my old one, but with the same mods folder, it still crashes, so it's nothing to do with the saved version.

So I can't tell if the 'SDL2_ttf.dll' problem is fixed (what this issue was about before I sidetracked it with the SDL mapping ;) ), but the change with the `.0` for `SDL2-2.0.so` is working, I didn't need to manually symlink the version without the `.0` anymore (but as said, I needed to delete the bundled shared objects for SDL).

The error I get when trying to load or create a project, is this (should I create a new issue for this to not mix up this more, or did something go wrong with my build?): 

```
Unhandled exception. System.IndexOutOfRangeException: Index was outside the bounds of the array.
   at YAFC.Parser.LuaContext.Exec(Byte[] chunk, Int32 length, String mod, String name, Int32 argument) in /home/benjamin/workspaces/misc/yafc/YAFCparser/LuaContext.cs:line 382
   at YAFC.Parser.LuaContext.Require(IntPtr lua) in /home/benjamin/workspaces/misc/yafc/YAFCparser/LuaContext.cs:line 344
   at YAFC.Parser.LuaContext.lua_pcallk(IntPtr state, Int32 nargs, Int32 nresults, Int32 msgh, IntPtr ctx, IntPtr k)
   at YAFC.Parser.LuaContext.Exec(Byte[] chunk, Int32 length, String mod, String name, Int32 argument) in /home/benjamin/workspaces/misc/yafc/YAFCparser/LuaContext.cs:line 406
   at YAFC.Parser.LuaContext.DoModFiles(String[] modorder, String fileName, IProgress`1 progress) in /home/benjamin/workspaces/misc/yafc/YAFCparser/LuaContext.cs:line 432
   at YAFC.Parser.FactorioDataSource.Parse(String factorioPath, String modPath, String projectPath, Boolean expensive, IProgress`1 progress, ErrorCollector errorCollector) in /home/benjamin/workspaces/misc/yafc/YAFCparser/FactorioDataSource.cs:line 238
   at YAFC.WelcomeScreen.LoadProject() in /home/benjamin/workspaces/misc/yafc/YAFC/Windows/WelcomeScreen.cs:line 150
Fatal error. Internal CLR error. (0x80131506)
   at YAFC.Parser.LuaContext.lua_pcallk(IntPtr, Int32, Int32, Int32, IntPtr, IntPtr)
   at YAFC.Parser.LuaContext.lua_pcallk(IntPtr, Int32, Int32, Int32, IntPtr, IntPtr)
   at YAFC.Parser.LuaContext.Exec(Byte[], Int32, System.String, System.String, Int32)
   at YAFC.Parser.LuaContext.DoModFiles(System.String[], System.String, System.IProgress`1<System.ValueTuple`2<System.String,System.String>>)
   at YAFC.Parser.FactorioDataSource.Parse(System.String, System.String, System.String, Boolean, System.IProgress`1<System.ValueTuple`2<System.String,System.String>>, YAFC.Model.ErrorCollector)
   at YAFC.WelcomeScreen+<LoadProject>d__21.MoveNext()
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[[System.Threading.Tasks.VoidTaskResult, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[YAFC.WelcomeScreen+<LoadProject>d__21, YAFC, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].ExecutionContextCallback(System.Object)
   at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[[System.Threading.Tasks.VoidTaskResult, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[YAFC.WelcomeScreen+<LoadProject>d__21, YAFC, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].MoveNext(System.Threading.Thread)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[[System.Threading.Tasks.VoidTaskResult, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[YAFC.WelcomeScreen+<LoadProject>d__21, YAFC, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].MoveNext()
   at YAFC.UI.EnterThreadPoolAwaitable+<>c.<.cctor>b__7_0(System.Object)
   at System.Threading.QueueUserWorkItemCallback+<>c.<.cctor>b__6_0(System.Threading.QueueUserWorkItemCallback)
   at System.Threading.ExecutionContext.RunForThreadPoolUnsafe[[System.__Canon, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]](System.Threading.ExecutionContext, System.Action`1<System.__Canon>, System.__Canon ByRef)
   at System.Threading.QueueUserWorkItemCallback.Execute()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
   at System.Threading._ThreadPoolWaitCallback.PerformWaitCallback()
```

**Comment by ShadowTheAge** [comment]
No, thats a different issue that is related to the code I wrote just yesterday, and havent fully tested it yet, but thanks for the finding!

**Comment by ShadowTheAge** [comment]
I have published 0.4.3 but after reviewing the latest stack trace I now have doubts that it is related to the (previous comment).
Please check if that issue still persists

**Comment by SuperTux88** [comment]
Looks like it's all fixed now. I can load my file without the stack trace above and I can also select text in input field without an error about `SDL2_ttf.dll`. And it finds all my shared objects without manual changes or symlinks.

(The only manual change that is needed now is that I need to run `chmod +x ./YAFC` before I can launch it with `./YAFC`, because it's missing the executable flag. But I guess that is because you build everything on Windows and I don't know if Windows knows something like an executable flag (I think not), when I built it myself yesterday, the executable flag was correctly set. But that's now really a minor thing, Linux users usually know how to set this (that's sometimes also needed for other manually downloaded stuff) and you could also run it with `dotnet YAFC.dll` as you mentioned it in the Linux instructions.)

So since the main bug here is fixed, I close this. Thanks for the fast fix :heart:
